# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/vr/ipc/VRLayerChild.cpp
# Commit: 0b191b707750
# Full Hash: 0b191b707750e040289033359306e4966c1bf062
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-06-15 21:48:38
# Regressor Bug: 1632249
# File Overlap Count: 1
# Description:
#   Bug 1632249 - WebXR compositing fixes. r=imanol
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D78117
# ==============================================================================

diff -r 16c5dd2cfd60 -r 0b191b707750 gfx/vr/ipc/VRLayerChild.cpp
--- a/gfx/vr/ipc/VRLayerChild.cpp	Mon Jun 15 18:25:55 2020 +0000
+++ b/gfx/vr/ipc/VRLayerChild.cpp	Mon Jun 15 18:25:58 2020 +0000
@@ -10,6 +10,7 @@
 #include "mozilla/layers/ImageBridgeChild.h"
 #include "mozilla/layers/LayersMessages.h"  // for TimedTexture
 #include "mozilla/layers/SyncObject.h"      // for SyncObjectClient
+#include "mozilla/StaticPrefs_webgl.h"
 
 #include "ClientWebGLContext.h"
 #include "gfxPlatform.h"
@@ -40,6 +41,7 @@
 void VRLayerChild::SetXRFramebuffer(WebGLFramebufferJS* fb) {
   mFramebuffer = fb;
 }
+
 static constexpr bool kIsAndroid =
 #if defined(MOZ_WIDGET_ANDROID)
     true;
@@ -65,7 +67,6 @@
   // 1 extra frame, accomodating overlapped asynchronous rendering.
   mLastFrameTextureDesc = mThisFrameTextureDesc;
 
-  bool forSeparateVrSwapChain = false;
   bool getNewFrame = true;
   if (kIsAndroid) {
     /**
@@ -75,17 +76,24 @@
      * in the WebGLContext GLScreenBuffer producer. Not doing so causes some
      * freezes, crashes or other undefined behaviour.
      */
-    forSeparateVrSwapChain = true;
     getNewFrame = (!mThisFrameTextureDesc ||
                    aDisplayInfo.mDisplayState.lastSubmittedFrameId ==
                        mLastSubmittedFrameId);
   }
   if (getNewFrame) {
-    webgl::Present(*webgl);
+    const RefPtr<layers::ImageBridgeChild> imageBridge =
+        layers::ImageBridgeChild::GetSingleton();
 
-    RefPtr<layers::ImageBridgeChild> imageBridge =
-        layers::ImageBridgeChild::GetSingleton();
-    mThisFrameTextureDesc = webgl::GetFrontBuffer(*webgl, imageBridge);
+    auto texType = layers::TextureType::Unknown;
+    if (imageBridge) {
+      texType = layers::PreferredCanvasTextureType(*imageBridge);
+    }
+    if (kIsAndroid && StaticPrefs::webgl_enable_surface_texture()) {
+      texType = layers::TextureType::AndroidNativeWindow;
+    }
+
+    webgl->Present(mFramebuffer, texType);
+    mThisFrameTextureDesc = webgl->GetFrontBuffer(mFramebuffer, texType);
   }
 
   mLastSubmittedFrameId = frameId;
